### ISSUE TITLE
Improve Site Performance with Lazy Loading and WEBP Images

### ISSUE THREAD

**Issue by shivamgupta2020** [knative/docs#5981]
**Describe the change you'd like to see**
Implement two changes to enhance site performance:
1. Add loading="lazy" to image tags to improve Cumulative Layout Shift (CLS).
2. Convert PNG/JPG/JPEG images to WEBP format for faster loading times.

**Additional context**
These optimizations will reduce initial page load time and bandwidth usage, improving user experience and SEO.

**Comment by aliok** [comment]
Hey @shivamgupta2020 

Thanks a lot for the suggestions.

I will have some comments...

> Add loading="lazy" to image tags to improve Cumulative Layout Shift (CLS).

Sounds good.

> Convert PNG/JPG/JPEG images to WEBP format for faster loading times.

We will need a bash script that iterates over the website content and converts images to webp. We can't do this manuallly.
This also means, we need another script that checks if images are converted to webp as a GitHub+Prow check and blocks merging the PR if there are images not converted.

However, let's talk about specifics later. We would need some data first, to see how much improvement there is.

We would need 4 data:
- What's the page size and page load time currently?
- Same metrics, but with `loading=lazy`
- Same metrics, but with `webp`
- Same metrics, but with both `loading=lazy` and `webp`

I don't expect to see these numbers for the entire website. Only the homepage is enough.

**Comment by shivamgupta2020** [comment]
@aliok 
Here are the metrics for all four of you requested.

**Metrics without `loading=lazy` and `webp`:-** 
![metrics without lazyloading   webp](https://github.com/knative/docs/assets/91847638/0bed8485-a82f-4fc8-bb6e-8358d98b3f09)
**Metrics wtth `webp`:-**
![metrics with webp](https://github.com/knative/docs/assets/91847638/4f9f8efb-9232-4f9b-bf94-16afb63fa508)
**Metrics with `loading=lazy`:-**
![metrics with lazy loading](https://github.com/knative/docs/assets/91847638/eb74d859-ac10-4e65-afbc-39db2a5a3f48)
**Metrics with `loading=lazy` and `webp`:-**
![metrics with lazy loading and webp](https://github.com/knative/docs/assets/91847638/7909db94-fefa-4ee9-bb82-687e106f1275)

**Comment by aliok** [comment]
@shivamgupta2020 what's the tool you've used to collect these numbers? I am curious.

**Comment by shivamgupta2020** [comment]
> @shivamgupta2020 what's the tool you've used to collect these numbers? I am curious.

I used Google Lighthouse.

**Comment by shivamgupta2020** [comment]
Hi @shivansh-bhatnagar18 , I've completed most of the work on this issue and am waiting for the maintainer's approval to raise a PR. You can contribute to other issues, so I'm unassigning you for now.
/unassign @shivansh-bhatnagar18

**Comment by aliok** [comment]
@shivamgupta2020 
I see there's not much improvement with webp. Let's not go that route.

Can you create a small POC for using `loading=lazy`? I wonder how we can do that with Markdown and MkDocs

**Comment by shivamgupta2020** [comment]
@aliok 
So, I looked into it online. You can make images load lazily by adding` {: loading=lazy }` to the image tag.
Reference:- 
1. https://github.com/squidfunk/mkdocs-material/issues/1867
2. https://squidfunk.github.io/mkdocs-material/reference/images/#image-alignment

**Comment by aliok** [comment]
Thanks @shivamgupta2020 

I think I would go this route...

I would create a PR with 2 things:
- Homepage images converted to lazy
- A script (in Python) that checks that page for any images without lazy loading and fails --> add a new GitHub action for this

**Comment by rohitsrma** [comment]
Hello @aliok, @shivamgupta2020!

I followed the discussion, and this enhancement looks really interesting to me. If there's any way I can help, especially with the Python scripting part, I'm more than willing to do so.

One aspect we need to consider is its impact on users with slower internet speeds, as it could cause layout shifting issues. For example, when I tested lazy loading on the first two images with a slow internet connection, there were noticeable layout shifts. While it's not a major issue, it's worth thinking about possible ways to mitigate it. What are your thoughts on this?

![Screenshot (61)](https://github.com/knative/docs/assets/111359305/79ed054c-ef40-42df-afa2-3af07c15d392)
